### PR TITLE
Add map support with mapProp decorator

### PR DESCRIPTION
**Pull request by k0hamed**
- Adding mapProp decorator
- Adding mapPtop tests
- Adding documentation for the new mapProp

### CODE REVIEW

File: src/prop.ts [hunk 1]
```diff
@@ -195,6 +206,18 @@ const baseProp = (rawOptions: any, Type: any, target: any, key: any, isArray = f
     return;
   }
 
+  if (isMap) {
+    schema[name][key] = {
+      ...schema[name][key],
+      type: Map,
+      ...options
+    };
+    schema[name][key].of = {
```
Comment (hasezoey) [inline src/prop.ts]: why 2 different "calls" and not combine it into one?

File: src/test/index.test.ts [hunk 2]
```diff
@@ -186,6 +187,34 @@ describe('Typegoose', () => {
...
+          link: 'url'
+        },
+        'day2': {
+          content: 'day2',
+          link: 'url//2'
+        },
+      },
+      projects: {},
+    });
+    expect(user).to.be.ok;
```
Comment (hasezoey) [inline src/test/index.test.ts]:
```suggestion
    expect(user).to.not.be.an('undefined');
```

File: src/test/index.test.ts [hunk 3]
```diff
@@ -186,6 +187,34 @@ describe('Typegoose', () => {
...
+    expect(user).to.have.property('socialNetworks').to.be.instanceOf(Map);
+    expect(user.socialNetworks.get('twitter')).to.be.eq('twitter account');
+    expect(user.socialNetworks.get('facebook')).to.be.eq('facebook account');
+    expect(user).to.have.property('sideNotes').to.be.instanceOf(Map);
+    expect(user.sideNotes.get('day1')).to.have.property('content', 'day1');
+    expect(user.sideNotes.get('day1')).to.have.property('link', 'url');
+    expect(user.sideNotes.has('day2')).to.be.true;
```
Comment (hasezoey) [inline src/test/index.test.ts]:
```suggestion
    expect(user.sideNotes.has('day2')).to.be.equal(true)
```

**Comment by hasezoey** [comment]
@k0hamed when you are still working with / on typegoose, could you rebase on master & squash some commits and resolve the requested changes above? (Otherwise i would takeover and make these in a seperate pr)

**Comment by k0hamed** [comment]
@hasezoey No problem, I will do that at the weekend.

**Comment by hasezoey** [comment]
@k0hamed sorry, no need to, already done it in #342, just had the time, thanks anyway ^^

btw you should look at your last message, to confirm nothing from you got leaked (you included some of the github email)

**Pull request closed without merging**